Updating "Testing and Debugging" article (#267)
MicrosoftDocs/quantum-docs: articles/user-guide/testing-debugging.md

# Testing and debugging

> As with classical programming, it is essential to be able to check that quantum programs act as intended, and to be able to diagnose incorrect behavior. Unlike with classical software development, though, observing the state of a quantum system and tracking the behavior of a quantum program might not be easy.

Reviewer: ```suggestion
As with classical programming, it is essential to be able to check that quantum programs act as intended, and to be able to diagnose incorrect behavior. Unlike classical programming, though, observing the state of a quantum system and tracking the behavior of a quantum program is not always easy.
```

```qsharp
namespace DumpOperationTest {

    open Microsoft.Quantum.Canon;
    open Microsoft.Quantum.Intrinsic;
    open Microsoft.Quantum.Diagnostics;
    
    @EntryPoint()
    operation DumpCnot() : Unit {
    DumpOperation(2, ApplyToFirstTwoQubitsCA(CNOT, _));
    }
}
```
Calling `DumpOperation` will print the following matrix,

```output
Real:
[[1, 0, 0, 0],
[0, 0, 0, 1],
[0, 0, 1, 0],
[0, 1, 0, 0]]
Imag:
[[0, 0, 0, 0],
[0, 0, 0, 0],
[0, 0, 0, 0],
[0, 0, 0, 0]]
```
> [!NOTE]
> `DumpOperation` and the rest of Dump functions use the little-endian encoding for converting basis states to the indices of matrix elements. Thus, the second column of the CNOT matrix corresponds to the input state |1⟩\_{LE} = |10⟩, which the CNOT gate converts to |11⟩ = |3\_{LE}. Similarly, the imput state |2⟩\_{LE} = |01⟩.

### DumpMachine

> The full-state quantum simulator distributed as part of the Quantum Development Kit writes into the file the [wave function](https://en.wikipedia.org/wiki/Wave_function) of the entire quantum system, as a one-dimensional array of complex numbers, in which each element represents the amplitude of the probability of measuring the computational basis state $\ket{n}$, where $\ket{n} = \ket{b_{n-1}...b_1b_0}$ for bits $\{b_i\}$. For example, consider the following operation that allocates two qubits and prepares an uneven superposition state on them.

Reviewer: ```suggestion
The full-state quantum simulator distributed as part of the Quantum Development Kit writes into the file the [wave function](https://en.wikipedia.org/wiki/Wave_function) of the entire quantum system, as a one-dimensional array of complex numbers, in which each element represents the amplitude of the probability of measuring the computational basis state $\ket{n}$, where $\ket{n} = \ket{b_{n-1}...b_1b_0}$ for bits $\{b_i\}$. For example, consider the following operation that allocates two qubits and prepares an uneven superposition state on them:
```

***

#### DumpMachine with Jupyter Notebook

> For the sake of simplicity, in the previous testing and debugging tools we have displayed examples of code using Q# standalone application in the command prompt and any IDE, though you can use any of the running options offered by Quantum Development Kit to develop quantum computing applications in Q#.

Reviewer: ```suggestion
For the sake of simplicity, in the previous testing and debugging tools we displayed examples of code using a Q# standalone application at a command prompt and any IDE. However, you can use any of the development options offered by Quantum Development Kit to develop quantum computing applications in Q#. For more information, see [Set up the QDK](xref:microsoft.quantum.install-qdk.overview).
```

> In this case, for <xref:Microsoft.Quantum.Diagnostics.DumpMachine>, we explicitly show the development on Q# Jupyter Notebook as it offers more visualization tools for testing and debugging quantum programs.

Reviewer: ```suggestion
In this example for <xref:Microsoft.Quantum.Diagnostics.DumpMachine>, we explicitly show the development using a Q# Jupyter Notebook, as it offers more visualization tools for testing and debugging quantum programs.
```

> 1. To run `DumpMachine` on Jupyter Notebook, open a [new Jupyter Notebook with a Q# kernel](xref:microsoft.quantum.install-qdk.overview.jupyter) and copy the following code to the first notebook cell.

Reviewer: ```suggestion
1. To run `DumpMachine` on a Jupyter Notebook, open a [new Jupyter Notebook with a Q# kernel](xref:microsoft.quantum.install-qdk.overview.jupyter) and copy the following code to the first notebook cell:
```

 ```qsharp
open Microsoft.Quantum.Diagnostics;
 
operation MultiQubitDumpMachineDemo() : Unit {
        use qubits = Qubit[2];
        X(qubits[1]);
        H(qubits[1]);
        R1Frac(1, 2, qubits[1]);
        
        DumpMachine();

        ResetAll(qubits);
    }

```

> 1. In a new cell, run `MultiQubitDumpMachineDemo` operation on a full state quantum simulator by using the `%simulate` magic command. `DumpMachine` call will print the information about the quantum state of the program after the Controlled Ry gate as a set of lines, one per basis state, showing their complex amplitudes, phases, and measurement probabilities:

Reviewer: ```suggestion
1. In a new cell, run the `MultiQubitDumpMachineDemo` operation on a full state quantum simulator by using the `%simulate` magic command. The `DumpMachine` call prints the information about the quantum state of the program after the Controlled Ry gate as a set of lines, one per basis state, showing their complex amplitudes, phases, and measurement probabilities.
```

![DumpMachine output](~/media/dumpmachine-output.png)

> [!NOTE]

> > You can use <xref:microsoft.quantum.iqsharp.magic-ref.config> (available only in Q# Jupyter Notebooks) to tweak the format of `DumpMachine` output. It offers a lot of settings convenient in different scenarios. For example, by default `DumpMachine` uses little-endian integers to denote the basis states (the first column of the output); if you find raw bit strings easier to read, you can use `%config dump.basisStateLabelingConvention="Bitstring"` to switch.

Reviewer: ```suggestion
> You can use <xref:microsoft.quantum.iqsharp.magic-ref.config> (available only in Q# Jupyter Notebooks) to tweak the format of the `DumpMachine` output. It offers many settings that you can use in different scenarios. For example, by default `DumpMachine` uses little-endian integers to denote the basis states (the first column of the output); if you find raw bit strings easier to read, you can use `%config dump.basisStateLabelingConvention="Bitstring"` to switch.
```

> 1. Jupyter Notebook offers the possibility to visualize the execution of the quantum program as a quantum circuit by using <xref:microsoft.quantum.iqsharp.magic-ref.trace> (available only in Q# Jupyter Notebooks). This command traces one run of the Q# program and build a circuit based on that execution. This is the circuit resulting from the running of `%trace MultiQubitDumpMachineDemo`, 

Reviewer: ```suggestion
1. Jupyter Notebooks offers the option to visualize the run of the quantum program as a quantum circuit by using <xref:microsoft.quantum.iqsharp.magic-ref.trace> (available only in Q# Jupyter Notebooks). This command traces one run of the Q# programs and build a circuit based on that run. This is the circuit resulting from the running of `%trace MultiQubitDumpMachineDemo`, 
```

![DumpMachine trace output](~/media/dumpmachine-trace-output.png)

   The visualization is interactive, allowing you to click on each block to drill down to the intrinsic gates.

> 1. Finally, <xref:microsoft.quantum.iqsharp.magic-ref.debug> (available only in Q# Jupyter Notebooks) allows you to combine tracing the program execution (as a circuit) and observing the program state as it evolves at the same time. The visualization is also interactive, you can click through each of the steps until the program run is complete, and switch to observe real and imaginary components of the amplitudes instead of measurement probabilities in the beginning of the program.

Reviewer: ```suggestion
1. Finally, <xref:microsoft.quantum.iqsharp.magic-ref.debug> (available only in Q# Jupyter Notebooks) allows you to combine tracing the program execution (as a circuit) and observing the program state as it evolves at the same time. The visualization is also interactive; you can click through each of the steps until the program run is complete, and switch to observe real and imaginary components of the amplitudes, instead of measurement probabilities in the beginning of the program.
```
